### PR TITLE
174 webgl context warning

### LINKED ISSUE #174: WebGL warning: Exceeded 16 live WebGL contexts for this principal, losing the least recently used one.
Hello, thanks for you good job :-)

I have this warning when I use gpu.js. I have this warning under firefox(v56) but not under chromium(v57) with the same code.

### PR DESCRIPTION
**Pull request by nickhingston**
add destroy method to kernel and gpu objects so we can clean up as much as humanly possible!

fixes #174

### CODE REVIEW

**Comment by nickhingston** [comment]
you might wanna review the commits separately...

File: test/issues/174-webgl-context-warning.js [hunk 1]
```diff
@@ -0,0 +1,77 @@
+(function() {
+
+	// unfortunately there is a bit of a manual process to check this properly!
```
[the remark this answers is not in the document]
Comment (nickhingston) [reply test/issues/174-webgl-context-warning.js]: yeah, fair point.  If you've got suggestions in this area though of how to do that would be helpful!  i think there might be an event we can listen to...

File: src/core/gpu.js [hunk 2]
```diff
@@ -382,6 +381,33 @@ class GPU extends GPUCore {
...
+		const destroyWebGl = !this._webGl && kernels.length && kernels[0]._webGl;
+		for (let i = 0; i < this.kernels.length; i++) {
+			this.kernels[i].destroy(true); // remove canvas if exists
+		}
+
+		if (destroyWebGl) {
+			destroyWebGl.OES_texture_float = null;
+			destroyWebGl.OES_texture_float_linear = null;
+			destroyWebGl.OES_element_index_uint = null;
+			destroyWebGl.getExtension('WEBGL_lose_context').loseContext();
```
[the remark this answers is not in the document]
Comment (nickhingston) [reply src/core/gpu.js]: i think so, but no harm in checking return value...

File: src/backend/web-gl/kernel.js [hunk 3]
```diff
@@ -190,11 +192,16 @@ module.exports = class WebGLKernel extends KernelBase {
 		const vertShader = gl.createShader(gl.VERTEX_SHADER);
 		gl.shaderSource(vertShader, compiledVertShaderString);
 		gl.compileShader(vertShader);
+		if (this.vertShader) {
```
Comment (robertleeplummerjr) [inline src/backend/web-gl/kernel.js]: Empty if.

File: test/issues/174-webgl-context-warning.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (robertleeplummerjr) [reply test/issues/174-webgl-context-warning.js]: We just would need to mock the WebGL api and test if the methods were called.  We could use a mocking utility recommended from: https://qunitjs.com/plugins/, such as sinon.  Then mocking would be like:

```js
const mockWebGL = {
  deleteTexture: sinon.spy(),
  deleteBuffer: sinon.spy(),
  deleteFramebuffer: sinon.spy(),
  deleteShader: sinon.spy(),
  deleteProgram: sinon.spy(),
};

const gpu = new GPU({ webGl: mockWebGL });

gpu.destroy();

QUnity.ok(mockWebGL.deleteTexture.called);
QUnity.ok(mockWebGL.deleteBuffer.called);
QUnity.ok(mockWebGL.deleteFramebuffer.called);
QUnity.ok(mockWebGL.deleteShader.called);
QUnity.ok(mockWebGL.deleteProgram.called);
```

**Comment by nickhingston** [comment]
as you were, safari doesn't support webgl2.

**Pull request merged**